src/sql/workbench/contrib/notebook/browser/cellToolbarActions.ts

```diff
@@ -0,0 +1,114 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Action } from 'vs/base/common/actions';
+import { localize } from 'vs/nls';
+import { INotebookService, INotebookEditor } from 'sql/workbench/services/notebook/browser/notebookService';
+import { NotebookComponent } from 'sql/workbench/contrib/notebook/browser/notebook.component';
+import { ICellModel } from 'sql/workbench/services/notebook/browser/models/modelInterfaces';
+import { CellType } from 'sql/workbench/services/notebook/common/contracts';
+import { ToggleableAction } from 'sql/workbench/contrib/notebook/browser/notebookActions';
+
+export class CellToolbarAction extends Action {
+
+	constructor(
+		id: string,
+		label: string,
+		cssClass: string,
+		tooltip: string,
+		private _cellModel: ICellModel,
+		@INotebookService private _notebookService: INotebookService
+	) {
+		super(id, label, cssClass);
+		this._tooltip = tooltip;
+	}
+	// Todo: Add actions references here?
+}
+
+export class AddCellAction extends Action {
+	public cellType: CellType;
+
+	constructor(
+		id: string, label: string, cssClass: string
+	) {
+		super(id, label, cssClass);
+	}
+	public async run(context: INotebookEditor): Promise<any> {
+		//Add Cell after current selected cell.
+		let index = 0;
+		if (context && context.cells) {
+			let notebookcomponent = context as NotebookComponent;
+			let id = notebookcomponent.activeCellId;
+			if (id) {
+				index = context.cells.findIndex(cell => cell.id === id);
+				index = index + 1;
+			}
+		}
+		context.addCell(this.cellType, index);
+	}
+}
+
+/**
+ * Not yet implemeneted
+ * Goal: show more actions available depending on what cell the toolbar is for.
+ * (ie. runAlActions when on code cell.)
+ */
+export class MoreActions extends Action {
+
+	constructor(
+		id: string, label: string, cssClass: string
+	) {
+		super(id, label, cssClass);
+	}
+}
+/**
+ * Todo: When a code cell is selected, the cursor focus is already set to the field within it. The below method needs to check for the active state first, then set the icon accordingly.
+ */
+export class EditCellAction extends ToggleableAction {
```

Comment:
This successfully toggles the icon. Is it appropriate to also toggle the edit mode and active/inactive state of the cell?

Reply:
Not sure what you mean - but yes I would expect that clicking the edit button would toggle the edit mode of the cell if that's what you mean.

Reply:
Yes, clicking the edit show activate edit more and clicking the toggled X icon should unselect the cell. I don't know how to get those working.